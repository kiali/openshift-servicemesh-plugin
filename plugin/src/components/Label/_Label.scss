/*
PF4 adds a wrapper for the label content
And the label color is override by .pf-c-label__content
And there are not styles with the color white
*/
.whiteColorLabelContent > .pf-c-label__content {
  color: #fff;
}
