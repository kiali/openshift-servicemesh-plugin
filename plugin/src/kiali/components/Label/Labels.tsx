import * as React from 'react';
import { Label } from './Label';
import { Button, ButtonVariant, Tooltip, TooltipPosition } from '@patternfly/react-core';
import { kialiStyle } from 'styles/StyleUtils';
import { KialiIcon } from '../../config/KialiIcon';

const SHOW_MORE_TRESHOLD = 2;

interface Props {
  labels?: { [key: string]: string };
  tooltipMessage?: string;
  expanded?: boolean;
}

interface State {
  expanded: boolean;
}

const linkStyle = kialiStyle({
  padding: '0 4px 0 4px',
  fontSize: '0.8rem',
  bottom: '2px'
});

const infoStyle = kialiStyle({
  margin: '4px 4px 2px 5px'
});

const labelsContainerStyle = kialiStyle({
  overflow: 'hidden'
});

export class Labels extends React.Component<Props, State> {
  constructor(props: Props, state: State) {
    super(props, state);
    this.state = {
      expanded: props.expanded ? props.expanded : false
    };
  }

  labelKeys() {
    return Object.keys(this.props.labels || {});
  }

  hasLabels() {
    return this.labelKeys().length > 0;
  }

  hasManyLabels() {
    return this.labelKeys().length > SHOW_MORE_TRESHOLD;
  }

  showItem(i: number) {
    return this.state.expanded || !this.hasManyLabels() || i < SHOW_MORE_TRESHOLD;
  }

  expandLabels = () => {
    this.setState({ expanded: true });
  };

  renderMoreLabelsLink() {
    if (this.hasManyLabels() && !this.state.expanded) {
      return (
        <Button
          data-test="label_more"
          key="label_more"
          variant={ButtonVariant.link}
          className={linkStyle}
          onClick={this.expandLabels}
        >
          More labels...
        </Button>
      );
    }

    return null;
  }

  renderLabels() {
    return this.labelKeys().map((key, i) => {
      return this.showItem(i) ? (
        <div key={'label_div_' + i} data-test={key + '-label-container'}>
          <Label key={'label_' + i} name={key} value={this.props.labels ? this.props.labels[key] : ''} />
        </div>
      ) : undefined;
    });
  }

  renderEmptyLabels() {
    return <span> No labels </span>;
  }

  render() {
    const tooltip = this.props.tooltipMessage ? (
      <Tooltip
        key={`tooltip_missing_sidecar`}
        position={TooltipPosition.auto}
        content={<div style={{ textAlign: 'left' }}>{this.props.tooltipMessage}</div>}
      >
        <KialiIcon.Info className={infoStyle} />
      </Tooltip>
    ) : undefined;
    return (
      <div className={labelsContainerStyle}>
        {this.hasLabels() ? [this.renderLabels(), this.renderMoreLabelsLink()] : this.renderEmptyLabels()}
        {tooltip}
      </div>
    );
  }
}
