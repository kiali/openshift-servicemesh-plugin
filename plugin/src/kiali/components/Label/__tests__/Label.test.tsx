import * as React from 'react';
import { shallow } from 'enzyme';
import { Label } from '../Label';
import { shallowToJson } from 'enzyme-to-json';

const mockBadge = (name = 'my_key', value = 'my_value') => {
  const component = <Label value={value} name={name} />;
  return shallow(component);
};

describe('#Badge render correctly with data', () => {
  it('should render badge', () => {
    const key = 'app';
    const value = 'bookinfo';
    const wrapper = mockBadge(key, value);

    expect(shallowToJson(wrapper)).toBeDefined();
    expect(shallowToJson(wrapper)).toMatchSnapshot();

    expect(wrapper.name()).toEqual('Label');

    const labelKey = wrapper.find('Label').getElements()[0];

    expect(labelKey.props.children).toEqual(`${key}=${value}`);
  });
});
