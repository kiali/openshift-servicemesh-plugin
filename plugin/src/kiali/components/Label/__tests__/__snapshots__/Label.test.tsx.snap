// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#Badge render correctly with data should render badge 1`] = `
<Label
  className="kiali_f1il4iwg"
  isCompact={true}
>
  app=bookinfo
</Label>
`;
